packages/validation/src/validation-controller.ts

```diff
@@ -0,0 +1,585 @@
+    if (propertyInfo === void 0) { return; }
+
+    const { object, propertyName } = propertyInfo;
+    await this.validate(new ValidateInstruction(object, propertyName, rules));
+  }
+
+  public resetBinding(binding: BindingWithBehavior) {
+    const bindingInfo = this.bindings.get(binding);
+    if (bindingInfo === void 0) { return; }
+
+    const propertyInfo = this.getPropertyInfo(binding, bindingInfo);
+    if (propertyInfo === void 0) { return; }
+
+    bindingInfo.propertyInfo = void 0;
+    const { object, propertyName } = propertyInfo;
+    this.reset(new ValidateInstruction(object, propertyName));
+  }
+
+  public async revalidateErrors() {
+    const map = this.results
+      .reduce(
+        (acc, { isManual, object, propertyRule, rule, valid }) => {
+          if (!valid && !isManual && propertyRule !== void 0 && object !== void 0 && rule !== void 0) {
+            let value = acc.get(object);
+            if (value === void 0) {
+              value = new Map();
+              acc.set(object, value);
+            }
+            let rules = value.get(propertyRule);
+            if (rules === void 0) {
+              rules = [];
+              value.set(propertyRule, rules);
+            }
+            rules.push(rule);
+          }
+          return acc;
+        },
+        new Map<IValidateable, Map<PropertyRule, BaseValidationRule[]>>());
+
+    const promises = [];
+    for (const [object, innerMap] of map) {
+      promises.push(
+        this.validate(new ValidateInstruction(
+          object,
+          undefined,
+          Array.from(innerMap)
+            .map(([
+              { validationRules, messageProvider, property },
+              rules
+            ]) => new PropertyRule(validationRules, messageProvider, property, [rules]))
+        ))
+      );
+    }
+    await Promise.all(promises);
+  }
+  // TODO: have dispose
+
+  /**
+   * Interprets the instruction and returns a predicate that will identify relevant results in the list of rendered validation results.
+   */
+  private getInstructionPredicate(instruction?: ValidateInstruction): ValidationPredicate {
+    if (instruction === void 0) { return () => true; }
+
+    const propertyName = instruction.propertyName;
+    const rules = instruction.rules;
+
+    return x => x.object === instruction.object &&
+      (propertyName === void 0 || x.propertyName === propertyName) &&
+      (rules === void 0 || rules.includes(x.propertyRule!) ||
+        rules.some((r) => x.propertyRule === void 0 || r.$rules.flat().every(($r) => x.propertyRule!.$rules.flat().includes($r)))
+      );
+  }
+
+  private getPropertyInfo(binding: BindingWithBehavior, info: BindingInfo, flags: LifecycleFlags = LifecycleFlags.none): PropertyInfo | undefined {
+    let propertyInfo = info.propertyInfo;
+    if (propertyInfo !== void 0) {
+      return propertyInfo;
+    }
+
+    const scope = info.scope;
+    let expression = binding.sourceExpression.expression;
```

Comment:
```suggestion
    let expression = binding.sourceExpression.expression as IsBindingBehavior;
```

That's a union of all basic types, so you don't need to cast each expression in the `switch` statement down below

Reply:
This is cool :) Changed it.